ocaml/xapi-idl/xen/device_number.ml

```diff
@@ -119,112 +106,119 @@ let rec string_of_int26 x =
     in
-    inner f [] x
+    b26 - 1
+  in
+
+  let parse_int x =
+    match String.Sub.span ~min:1 ~sat:Char.Ascii.is_digit x with
+    | i, s ->
+        Option.map (fun i -> (i, s)) (String.Sub.to_int i)
+  in
+  let parse_b26 x =
+    match String.Sub.span ~min:1 ~sat:Char.Ascii.is_lower x with
+    | b, s ->
+        (b26_to_int b, s)
   in
   (* Parse a string "abc123" into x, y where x is "abc" interpreted as base-26
      and y is 123 *)
   let parse_b26_int x =
-    let d, p = take letter x in
-    let d' = int26_of_string (String.implode d) in
-    let p' = if p = [] then 0 else int_of_string (String.implode p) in
-    (d', p')
+    let pre, x = parse_b26 x in
+    if String.Sub.is_empty x then
+      Some (pre, 0)
+    else
+      let* post, x = parse_int x in
+      if not (String.Sub.is_empty x) then
+        None
+      else
+        Some (pre, post)
   in
   (* Parse a string "123p456" into x, y where x = 123 and y = 456 *)
   let parse_int_p_int x =
-    let d, rest = take digit x in
-    match rest with
-    | 'p' :: rest ->
-        let p, _ = take digit rest in
-        (int_of_string (String.implode d), int_of_string (String.implode p))
-    | [] ->
-        (int_of_string (String.implode d), 0)
-    | _ ->
-        failwith
-          (Printf.sprintf "expected digit+ p digit+ got: %s" (String.implode x))
+    let parse_p x =
+      match String.Sub.head x with
+      | Some 'p' ->
+          Some (String.Sub.tail x)
+      | Some _ | None ->
+          None
+    in
+    let* pre, x = parse_int x in
+    if String.Sub.is_empty x then
+      Some (pre, 0)
+    else
+      let* x = parse_p x in
+      let* post, x = parse_int x in
+      if not (String.Sub.is_empty x) then
+        None
+      else
+        Some (pre, post)
   in
-  match String.explode x with
-  | 'x' :: 'v' :: 'd' :: rest ->
-      let disk, partition = parse_b26_int rest in
-      (Xen, disk, partition)
-  | 's' :: 'd' :: rest ->
-      let disk, partition = parse_b26_int rest in
-      (Scsi, disk, partition)
-  | 'f' :: 'd' :: rest ->
-      let disk, partition = parse_b26_int rest in
-      (Floppy, disk, partition)
-  | 'h' :: 'd' :: rest ->
-      let disk, partition = parse_b26_int rest in
-      (Ide, disk, partition)
-  | 'd' :: rest ->
-      let disk, partition = parse_int_p_int rest in
-      (Xen, disk, partition)
-  | _ ->
-      failwith (Printf.sprintf "Failed to parse device name: %s" x)
+  if String.is_prefix ~affix:"xvd" x then
+    let rest = String.sub_with_range ~first:3 x in
+    let* disk, partition = parse_b26_int rest in
+    Some (Xen, disk, partition)
+  else if String.is_prefix ~affix:"sd" x then
+    let rest = String.sub_with_range ~first:2 x in
+    let* disk, partition = parse_b26_int rest in
+    Some (Scsi, disk, partition)
+  else if String.is_prefix ~affix:"fd" x then
+    let rest = String.sub_with_range ~first:2 x in
+    let* disk, partition = parse_b26_int rest in
+    Some (Floppy, disk, partition)
```

Comment:
This doesn't restrict the value to fda and fdb like `make` does, the code needs to call it